src/Sylius/Bundle/CoreBundle/StateMachineCallback/OrderPaymentCallback.php

```diff
@@ -42,22 +43,16 @@ public function updateOrderOnPayment(PaymentInterface $payment)
             throw new \RuntimeException(sprintf('Cannot retrieve Order from Payment with id %s', $payment->getId()));
         }
 
-        $total = 0;
-        if (PaymentInterface::STATE_COMPLETED === $payment->getState()) {
-            $payments = $order->getPayments()->filter(function (PaymentInterface $payment) {
-                return PaymentInterface::STATE_COMPLETED === $payment->getState();
-            });
+        $payments = $order->getPayments()->filter(function (PaymentInterface $payment) {
+            return PaymentInterface::STATE_COMPLETED === $payment->getState();
+        });
 
-            if ($payments->count() === $order->getPayments()->count()) {
-                $order->setPaymentState(PaymentInterface::STATE_COMPLETED);
-            }
-
-            $total += $payment->getAmount();
-        } else {
-            $order->setPaymentState($payment->getState());
+        $completedPaymentTotal = 0;
+        foreach ($payments as $payment) {
+            $completedPaymentTotal += $payment->getAmount();
         }
 
-        if ($total >= $order->getTotal()) {
+        if ($completedPaymentTotal >= $order->getTotal()) {
             $this->factory->get($order, OrderTransitions::GRAPH)->apply(OrderTransitions::TRANSITION_FULFILL, true);
```

Comment:
Here we should apply transition to order payment state machine and mark it as `PAID`. Logic for fulfilling order is not yet specified.